File: packages/azure/ios/AzureSession.swift
import AVFoundation
import Foundation
import MicrosoftCognitiveServicesSpeech

/// Drives an Azure SDK recognizer in continuous mode (speech-recognition by
/// default, conversation-transcriber when `config.enableSpeakerDiarization` is
/// true) and surfaces partials/finals/errors via `AzureCallbacks`.
/// Lifecycle: prepare → start → (events) → stop/discard → release.
internal final class AzureSession {
  private let config: AzureConfig
  private weak var callbacks: AzureCallbacks?

  private var speechConfig: SPXSpeechConfiguration?
  private var engine: AzureRecognizerEngine?
  private var audioMonitor: AudioSessionMonitor?

  /// Phrase finals accumulated during the session — stitched together on stop().
  private var accumulatedFinals: [String] = []
  /// Most recent partial — fallback when stop() lands before any final.
  private var latestPartial: String = ""

  private var sessionId: String = ""
  private var sessionStartMs: Double = 0

  /// True between start() and stop/discard/release.
  private var active: Bool = false
  /// While > Date.now, accept tail Recognized events that arrive after stop()
  /// returned, and emit them as additional 'final' callbacks. Lets the UI feel
  /// snappy on Stop while still surfacing any audio Azure was mid-processing.
  private var tailWindowEnd: Date?

  private var released: Bool = false

  /// How long after stop() we keep listening for tail Recognized events.
  private static let tailWindowSeconds: TimeInterval = 2.0

  init(config: AzureConfig, callbacks: AzureCallbacks) {
    self.config = config
    self.callbacks = callbacks
  }

  // MARK: Lifecycle

  func prepare() async throws {
    if released { throw AzureError(code: "ENGINE_NOT_READY", message: "Session was released") }

    try validateAuthConfig()
    try await ensureMicrophoneAuthorization()
    try configureAudioSession()
    installAudioMonitor()

    do {
      let cfg = try buildSpeechConfig()
      self.speechConfig = cfg

      // Wrap engine instantiation in an Obj-C @try/@catch — the SDK raises
      // raw NSException (not NSError) on init failures like requesting
      // diarization on a tier that doesn't include it. Without the catcher
      // those would crash the app instead of bubbling as a Clarion error.
      var built: AzureRecognizerEngine?
      var swiftError: Error?
      do {
        try AzureExceptionCatcher.run {
          do {
            built = self.config.enableSpeakerDiarization
              ? try AzureConversationTranscriberEngine(speechConfig: cfg, callbacks: self)
              : try AzureSpeechRecognizerEngine(speechConfig: cfg, callbacks: self)
          } catch {
            swiftError = error
          }
        }
      } catch {
        // Diarization fallback: if degradeOnTierMismatch is on, try a regular
        // speech recognizer instead, and emit a DEGRADED_MODE warning.
        if self.config.enableSpeakerDiarization && self.config.degradeOnTierMismatch {
          NSLog("[ClarionAzure] Diarization init failed; falling back to non-diarization.")
          do {
            built = try AzureSpeechRecognizerEngine(speechConfig: cfg, callbacks: self)
            self.callbacks?.onError(AzureError(
              code: "INTERNAL_ERROR",
              message: "Diarization unavailable on this Azure tier — fell back to non-diarization recognition. (DEGRADED_MODE)",
              recoverable: false
            ))
          } catch {
            throw mapEngineInitError(error as NSError)
          }
        } else {
          throw mapEngineInitError(error as NSError)
        }
      }
      if let swiftError = swiftError {
        throw AzureErrorMapping.map(swiftError)
      }
      guard let engine = built else {
        throw AzureError(code: "INTERNAL_ERROR", message: "Failed to initialize Azure recognizer")
      }
      self.engine = engine

      // Install custom-vocab phrase hints (no-op if list is empty).
      installPhraseHints()

      // NOTE: Pre-warming via SPXConnection.open() was attempted here, but the
      // Microsoft Speech SDK can raise SIGTRAP (C++ abort, not NSException) on
      // some configurations — and SIGTRAP cannot be caught by Obj-C @try/@catch
      // or Swift `try`. The first `start()` carries the ~500-1000 ms handshake
      // instead, which is acceptable. If the SDK exposes a safe pre-warm hook
      // in a future version, revisit.
    } catch let azureErr as AzureError {
      throw azureErr
    } catch {
      throw AzureErrorMapping.map(error)
    }
  }

  /// Maps an engine-init NSError (from the Obj-C catcher) into a friendly
  /// AzureError. Special-cases the common diarization-tier-mismatch.
  private func mapEngineInitError(_ err: NSError) -> AzureError {
    let reason = err.localizedDescription
    if config.enableSpeakerDiarization {
      return AzureError(
        code: "UNSUPPORTED_FORMAT",
        message: "Failed to initialize Azure conversation transcriber for diarization. "
          + "This feature requires the S0 (Standard) Speech tier and an en-US locale. "
          + "Underlying error: \(reason)"
      )
    }
    return AzureError(
      code: "INTERNAL_ERROR",
      message: "Failed to initialize Azure recognizer: \(reason)"
    )
  }

  func start() throws {
    if released { throw AzureError(code: "ENGINE_NOT_READY", message: "Session was released") }
    guard let engine = engine else {
      throw AzureError(code: "ENGINE_NOT_READY", message: "Call prepare() first")
    }

    accumulatedFinals.removeAll()
    latestPartial = ""
    sessionId = UUID().uuidString
    sessionStartMs = Date().timeIntervalSince1970 * 1_000
    tailWindowEnd = nil  // close any prior tail window — we're a new session

    do {
      try engine.start()
    } catch {
      throw AzureErrorMapping.map(error)
    }

    active = true
    callbacks?.onState("recording")
  }

  /// Optimistic stop: resolves immediately with the stitched session-final, then
  /// keeps the recognizer alive for `tailWindowSeconds` so any tail Recognized
  /// events Azure emits during its buffer flush surface as additional 'final'
  /// callbacks (the app can update its UI with the more-complete transcript).
  func stop() throws -> AzureTranscript {
    if released { throw AzureError(code: "ENGINE_NOT_READY", message: "Session was released") }
    guard let engine = engine else {
      throw AzureError(code: "ENGINE_NOT_READY", message: "No active recognition")
    }

    active = false
    tailWindowEnd = Date().addingTimeInterval(AzureSession.tailWindowSeconds)
    callbacks?.onState("stopping")

    // Kick off the SDK stop — don't await. The recognizer will fire any tail
    // Recognized events on the listener thread; sessionStopped lands a moment
    // later and we'll observe it via azureSessionStopped().
    do {
      try engine.stop()
    } catch {
      // Don't fail stop() on SDK quirks during flush — we're resolving optimistically.
      // The error will surface (if real) via the `canceled` event handler.
    }

    let optimisticFinal = buildSessionFinal()
    callbacks?.onState("idle")
    return optimisticFinal
  }

  func discard() {
    active = false
    tailWindowEnd = nil
    if let engine = engine {
      try? engine.stop()
    }
    callbacks?.onState("idle")
  }

  /// Hot-swap the auth token on the currently-active session. The Azure SDK
  /// reads the token off the SPXRecognizer.properties bag on the next
  /// reconnect; the in-flight session is unaffected.
  func updateAuthToken(_ token: String) {
    speechConfig?.authorizationToken = token
    // Propagate to the live recognizer's property bag so the swap takes effect
    // on the next websocket reconnect. The base `SPXRecognizer` exposes
    // `properties` even though it doesn't have a direct `authorizationToken`
    // accessor like the subclasses do.
    engine?.spxBase.properties?.setPropertyTo?(token, by: .speechServiceAuthorizationToken)
  }

  func release() async {
    if released { return }
    released = true
    discard()
    engine = nil
    speechConfig = nil
    audioMonitor = nil  // observers cleaned up in deinit
    try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
    callbacks?.onState("released")
  }

  /// Listens for AVAudioSession interruption + route change events so we can
  /// surface them as clean ClarionErrors instead of leaving the session in an
  /// inconsistent state. The monitor calls back on the main queue.
  private func installAudioMonitor() {
    audioMonitor = AudioSessionMonitor(
      onInterruptionBegan: { [weak self] in
        guard let self, self.active else { return }
        self.callbacks?.onError(AzureError(
          code: "AUDIO_SESSION_INTERRUPTED",
          message: "Audio session interrupted by another app (phone call, Siri, etc.).",
          recoverable: true
        ))
        // Don't auto-discard — leave it to the JS layer to decide whether to
        // call stop() or wait for .ended and resume.
      },
      onInterruptionEnded: { [weak self] _ in
        // Caller can decide to resume by calling start() again. We don't
        // auto-resume because the audio engine state is uncertain.
        _ = self
      },
      onRouteChanged: { [weak self] reason in
        guard let self, self.active else { return }
        // Only surface meaningful route changes; ignore .categoryChange noise.
        switch reason {
        case .oldDeviceUnavailable, .newDeviceAvailable, .override:
          self.callbacks?.onError(AzureError(
            code: "AUDIO_ROUTE_CHANGED",
            message: "Audio route changed (reason \(reason.rawValue)).",
            recoverable: true
          ))
        default:
          return
        }
      }
    )
  }

  // MARK: Helpers

  /// Returns true if we should still surface a result, based on session state
  /// + tail window. Closes the tail window when it expires.
  private func acceptingEvents() -> Bool {
    if active { return true }
    if let deadline = tailWindowEnd, Date() < deadline {
      return true
    }
    tailWindowEnd = nil
    return false
  }

  private func buildSessionFinal() -> AzureTranscript {
    let text: String
    if !accumulatedFinals.isEmpty {
      text = accumulatedFinals.joined(separator: " ")
    } else {
      text = latestPartial
    }
    return AzureTranscriptBuilder.fromText(
      text,
      isFinal: true,
      sessionId: sessionId,
      sessionStartMs: sessionStartMs,
      language: config.language
    )
  }

  // MARK: Setup

  private func validateAuthConfig() throws {
    let hasKey = !config.subscriptionKey.isEmpty && !config.region.isEmpty
    let hasToken = !config.authToken.isEmpty && !config.region.isEmpty
    let hasEndpoint = !config.endpoint.isEmpty
    if !(hasKey || hasToken || hasEndpoint) {
      throw AzureError(
        code: "INVALID_STATE",
        message: "Azure config needs one of: subscriptionKey+region, authToken+region, or endpoint."
      )
    }
  }

  private func buildSpeechConfig() throws -> SPXSpeechConfiguration {
    let cfg: SPXSpeechConfiguration
    if !config.endpoint.isEmpty {
      guard let endpointCfg = try? SPXSpeechConfiguration(endpoint: config.endpoint) else {
        throw AzureError(code: "INVALID_STATE", message: "Failed to build SPXSpeechConfiguration from endpoint.")
      }
      cfg = endpointCfg
      if !config.subscriptionKey.isEmpty {
        cfg.setPropertyTo(config.subscriptionKey, by: .speechServiceConnectionKey)
      }
      if !config.authToken.isEmpty {
        cfg.authorizationToken = config.authToken
      }
    } else if !config.authToken.isEmpty {
      guard let tokenCfg = try? SPXSpeechConfiguration(authorizationToken: config.authToken, region: config.region) else {
        throw AzureError(code: "AUTH_FAILED", message: "Failed to build SPXSpeechConfiguration from auth token.")
      }
      cfg = tokenCfg
    } else {
      guard let keyCfg = try? SPXSpeechConfiguration(subscription: config.subscriptionKey, region: config.region) else {
        throw AzureError(code: "AUTH_FAILED", message: "Failed to build SPXSpeechConfiguration from subscription key.")
      }
      cfg = keyCfg
    }

    cfg.speechRecognitionLanguage = config.language

    // Profanity policy.
    switch config.profanity.lowercased() {
    case "removed": cfg.setProfanityOptionTo(SPXSpeechConfigProfanityOption.profanityRemoved)
    case "raw":     cfg.setProfanityOptionTo(SPXSpeechConfigProfanityOption.profanityRaw)
    case "none":    cfg.setProfanityOptionTo(SPXSpeechConfigProfanityOption.profanityRaw)
    default:        cfg.setProfanityOptionTo(SPXSpeechConfigProfanityOption.profanityMasked)
    }

    if config.outputFormat.lowercased() == "detailed" {
      cfg.outputFormat = .detailed
      cfg.requestWordLevelTimestamps()
    }

    // Silence-detection: auto-stop after N ms of silence (server-side VAD).
    // Property API used because the typed Swift property doesn't exist in
    // the iOS SDK — only the constant string identifier is reliable.
    if config.silenceTimeoutMs > 0 {
      cfg.setPropertyTo(
        "\(Int(config.silenceTimeoutMs))",
        by: .speechServiceConnectionEndSilenceTimeoutMs
      )
    }

    return cfg
  }

  /// Install phrase-list grammar on the recognizer if any hints were supplied.
  /// Must be called AFTER the recognizer instance exists, since the grammar
  /// is created from the recognizer, not the speech config.
  private func installPhraseHints() {
    guard !config.phraseHints.isEmpty, let engine = engine else { return }
    guard let grammar = SPXPhraseListGrammar(recognizer: engine.spxBase) else { return }
    for phrase in config.phraseHints where !phrase.isEmpty {
      grammar.addPhrase(phrase)
    }
  }

  // MARK: Audio session + permissions

  private func configureAudioSession() throws {
    let session = AVAudioSession.sharedInstance()
    var options: AVAudioSession.CategoryOptions = [.duckOthers]
    if #available(iOS 18.0, *) {
      options.insert(.allowBluetoothHFP)
    } else {
      options.insert(.allowBluetooth)
    }
    do {
      try session.setCategory(.playAndRecord, mode: .measurement, options: options)
      try session.setActive(true, options: .notifyOthersOnDeactivation)
    } catch {
      throw AzureError(code: "AUDIO_BUSY", message: "Failed to activate audio session: \(error.localizedDescription)")
    }
  }

  private func ensureMicrophoneAuthorization() async throws {
    if #available(iOS 17.0, *) {
      switch AVAudioApplication.shared.recordPermission {
      case .granted: return
      case .denied:
        throw AzureError(code: "PERMISSION_DENIED", message: "Microphone permission denied")
      default:
        let granted = await withCheckedContinuation { (cont: CheckedContinuation<Bool, Never>) in
          AVAudioApplication.requestRecordPermission { cont.resume(returning: $0) }
        }
        if !granted {
          throw AzureError(code: "PERMISSION_DENIED", message: "Microphone permission denied")
        }
      }
      return
    }
    let session = AVAudioSession.sharedInstance()
    switch session.recordPermission {
    case .granted: return
    case .denied:
      throw AzureError(code: "PERMISSION_DENIED", message: "Microphone permission denied")
    default:
      let granted = await withCheckedContinuation { (cont: CheckedContinuation<Bool, Never>) in
        session.requestRecordPermission { cont.resume(returning: $0) }
      }
      if !granted {
        throw AzureError(code: "PERMISSION_DENIED", message: "Microphone permission denied")
      }
    }
  }
}

// MARK: - AzureRecognizerCallbacks

extension AzureSession: AzureRecognizerCallbacks {
  func azureRecognizing(result: SPXRecognitionResult, speakerId: String) {
    if !active { return }  // Don't emit partials in the tail window.
    let transcript = AzureTranscriptBuilder.from(
      result: result,
      isFinal: false,
      sessionId: sessionId,
      sessionStartMs: sessionStartMs,
      fallbackLanguage: config.language,
      speakerId: speakerId
    )
    latestPartial = transcript.text
    if config.emitPartials {
      callbacks?.onPartial(transcript)
    }
  }

  func azureRecognized(result: SPXRecognitionResult, speakerId: String) {
    if !acceptingEvents() { return }
    guard result.reason == .recognizedSpeech else { return }
    let transcript = AzureTranscriptBuilder.from(
      result: result,
      isFinal: true,
      sessionId: sessionId,
      sessionStartMs: sessionStartMs,
      fallbackLanguage: config.language,
      speakerId: speakerId
    )
    if !transcript.text.isEmpty {
      accumulatedFinals.append(transcript.text)
    }
    callbacks?.onFinal(transcript)
  }

  func azureCanceled(
    reason: SPXCancellationReason,
    errorCode: SPXCancellationErrorCode,
    errorDetails: String?
  ) {
    // Graceful stop = end of stream. Not an error.
    if reason == .endOfStream { return }
    // Tail event after stop window closed — drop silently.
    if !acceptingEvents() { return }

    let mapped = AzureErrorMapping.map(
      reason: reason,
      errorCode: errorCode,
      errorDetails: errorDetails
    )
    callbacks?.onError(mapped)
  }

  func azureSessionStopped() {
    // Close the tail window — recognizer has finished flushing.
    tailWindowEnd = nil
  }

  func azureSpeechBoundary(kind: String, offsetMs: Double) {
    if !active { return }
    callbacks?.onSpeechBoundary(kind: kind, offsetMs: offsetMs)
  }
}

File: packages/azure/android/src/main/java/com/clarionhq/azure/AzureSession.kt
package com.clarionhq.azure

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import com.microsoft.cognitiveservices.speech.CancellationErrorCode
import com.microsoft.cognitiveservices.speech.CancellationReason
import com.microsoft.cognitiveservices.speech.OutputFormat
import com.microsoft.cognitiveservices.speech.PhraseListGrammar
import com.microsoft.cognitiveservices.speech.ProfanityOption
import com.microsoft.cognitiveservices.speech.PropertyId
import com.microsoft.cognitiveservices.speech.RecognitionResult
import com.microsoft.cognitiveservices.speech.ResultReason
import com.microsoft.cognitiveservices.speech.SpeechConfig
import com.microsoft.cognitiveservices.speech.audio.AudioConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.URI
import java.util.UUID

/**
 * Drives an Azure SDK recognizer in continuous mode (speech-recognition by
 * default, conversation-transcriber when `config.enableSpeakerDiarization` is
 * true) and surfaces partials/finals/errors via [AzureCallbacks].
 * Lifecycle: prepare → start → (events) → stop/discard → release.
 */
internal class AzureSession(
  private val context: Context,
  private val config: AzureConfig,
  private val callbacks: AzureCallbacks,
) : AzureRecognizerCallbacks {

  private var speechConfig: SpeechConfig? = null
  private var audioConfig: AudioConfig? = null
  private var engine: AzureRecognizerEngine? = null
  private var focusMonitor: AudioFocusMonitor? = null

  /** Phrase finals accumulated during the session — stitched together on stop(). */
  private val accumulatedFinals = mutableListOf<String>()
  /** Latest partial — fallback when stop() lands before any final. */
  @Volatile private var latestPartial: String = ""

  private var sessionId: String = ""
  private var sessionStartMs: Double = 0.0

  /** True between start() and stop/discard/release. */
  @Volatile private var active: Boolean = false
  /**
   * Epoch-ms until which we keep accepting tail Recognized events that arrive
   * after [stop] returned. Zero/past = closed.
   */
  @Volatile private var tailWindowEndMs: Long = 0L
  @Volatile private var released: Boolean = false

  companion object {
    /** How long after stop() we keep listening for tail Recognized events. */
    private const val TAIL_WINDOW_MS: Long = 2_000L
  }

  // MARK: Lifecycle

  suspend fun prepare() {
    if (released) throw AzureError("ENGINE_NOT_READY", "Session was released")

    ensureMicPermission()
    validateAuthConfig()
    installFocusMonitor()

    try {
      val cfg = buildSpeechConfig()
      val audio = AudioConfig.fromDefaultMicrophoneInput()
      val recEngine: AzureRecognizerEngine = buildEngine(cfg, audio)
      this.speechConfig = cfg
      this.audioConfig = audio
      this.engine = recEngine

      // Install custom-vocab phrase hints (no-op if list is empty).
      installPhraseHints()

      // NOTE: WebSocket pre-warm via `Connection.openConnection(true)` was
      // removed for parity with iOS — the equivalent SDK call there crashes
      // (SIGTRAP) on some configurations. The first `start()` pays the
      // ~500-1000 ms handshake instead. The JS-side auth pre-flight in
      // AzureEngine.prepare() already catches bad keys/regions early.
    } catch (e: AzureError) {
      throw e
    } catch (t: Throwable) {
      throw AzureErrorMapping.map(t)
    }
  }

  /**
   * Pick the right SDK class based on `enableSpeakerDiarization`. If the
   * diarization path throws and `degradeOnTierMismatch` is on, fall back to
   * the regular SpeechRecognizer and emit an informational error so the JS
   * layer can show a DEGRADED_MODE notice.
   */
  private fun buildEngine(cfg: SpeechConfig, audio: AudioConfig): AzureRecognizerEngine {
    if (!config.enableSpeakerDiarization) {
      return AzureSpeechRecognizerEngineImpl(cfg, audio, this)
    }
    return try {
      AzureConversationTranscriberEngineImpl(cfg, audio, this)
    } catch (t: Throwable) {
      if (!config.degradeOnTierMismatch) throw t
      callbacks.onError(
        AzureError(
          code = "INTERNAL_ERROR",
          message = "Diarization unavailable on this Azure tier — fell back to non-diarization recognition. (DEGRADED_MODE)",
        ),
      )
      AzureSpeechRecognizerEngineImpl(cfg, audio, this)
    }
  }

  suspend fun start() {
    if (released) throw AzureError("ENGINE_NOT_READY", "Session was released")
    val rec = engine ?: throw AzureError("ENGINE_NOT_READY", "Call prepare() first")

    accumulatedFinals.clear()
    latestPartial = ""
    sessionId = UUID.randomUUID().toString()
    sessionStartMs = System.currentTimeMillis().toDouble()
    tailWindowEndMs = 0L  // close any prior tail window — fresh session

    try {
      withContext(Dispatchers.IO) { rec.startAsync().get() }
    } catch (t: Throwable) {
      throw AzureErrorMapping.map(t)
    }

    active = true
    callbacks.onState("recording")
  }

  /**
   * Optimistic stop: returns immediately with the stitched session-final, then
   * keeps the recognizer alive for [TAIL_WINDOW_MS] so any tail Recognized
   * events Azure emits during its buffer flush surface as additional 'final'
   * callbacks.
   */
  suspend fun stop(): AzureTranscript {
    if (released) throw AzureError("ENGINE_NOT_READY", "Session was released")
    val rec = engine ?: throw AzureError("ENGINE_NOT_READY", "No active recognition")

    active = false
    tailWindowEndMs = System.currentTimeMillis() + TAIL_WINDOW_MS
    callbacks.onState("stopping")

    // Kick off the SDK stop — don't await. Tail events arrive on the listener
    // thread; sessionStopped lands a moment later and closes the window.
    runCatching {
      withContext(Dispatchers.IO) { rec.stopAsync() }
    }

    val optimisticFinal = buildSessionFinal()
    callbacks.onState("idle")
    return optimisticFinal
  }

  suspend fun discard() {
    active = false
    tailWindowEndMs = 0L
    val rec = engine
    if (rec != null) {
      runCatching {
        withContext(Dispatchers.IO) { rec.stopAsync() }
      }
    }
    callbacks.onState("idle")
  }

  /**
   * Hot-swap the auth token on the live recognizer. The SDK applies this on
   * the next websocket reconnect. Existing in-flight session is unaffected.
   * Delegates to the engine wrapper which knows which concrete subclass it
   * is and can reach its `properties` collection.
   */
  fun updateAuthToken(token: String) {
    speechConfig?.authorizationToken = token
    engine?.updateAuthToken(token)
  }

  suspend fun release() {
    if (released) return
    released = true
    discard()
    engine?.close()
    audioConfig?.close()
    speechConfig?.close()
    focusMonitor?.release()
    engine = null
    audioConfig = null
    speechConfig = null
    focusMonitor = null
    callbacks.onState("released")
  }

  /**
   * Request audio focus. Surface focus loss as a clean AUDIO_SESSION_INTERRUPTED
   * error instead of letting the session silently continue with no audio.
   */
  private fun installFocusMonitor() {
    if (focusMonitor != null) return
    val m = AudioFocusMonitor(
      context = context,
      onFocusLost = { transient ->
        if (!active) return@AudioFocusMonitor
        callbacks.onError(
          AzureError(
            code = "AUDIO_SESSION_INTERRUPTED",
            message = if (transient)
              "Audio focus lost transiently (another app, notification)."
            else
              "Audio focus permanently lost — recognition stopped.",
            recoverable = transient,
          ),
        )
      },
      onFocusRegained = { /* JS layer decides whether to retry */ },
    )
    m.acquire()
    focusMonitor = m
  }

  // MARK: Helpers

  /** True if we should still surface a result, based on session state + tail window. */
  private fun acceptingEvents(): Boolean {
    if (active) return true
    return System.currentTimeMillis() < tailWindowEndMs
  }

  private fun buildSessionFinal(): AzureTranscript = AzureTranscriptBuilder.fromText(
    text = accumulatedFinals.joinToString(" ").ifEmpty { latestPartial },
    isFinal = true,
    sessionId = sessionId,
    sessionStartMs = sessionStartMs,
    language = config.language,
  )

  // MARK: Setup

  private fun ensureMicPermission() {
    val granted = ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) ==
      PackageManager.PERMISSION_GRANTED
    if (!granted) {
      throw AzureError("PERMISSION_DENIED", "Microphone permission not granted")
    }
  }

  private fun validateAuthConfig() {
    val hasKey = config.subscriptionKey.isNotEmpty() && config.region.isNotEmpty()
    val hasToken = config.authToken.isNotEmpty() && config.region.isNotEmpty()
    val hasEndpoint = config.endpoint.isNotEmpty()
    if (!hasKey && !hasToken && !hasEndpoint) {
      throw AzureError(
        "INVALID_STATE",
        "Azure config needs one of: subscriptionKey+region, authToken+region, or endpoint.",
      )
    }
  }

  private fun buildSpeechConfig(): SpeechConfig {
    val cfg: SpeechConfig = when {
      config.endpoint.isNotEmpty() -> {
        val uri = URI.create(config.endpoint)
        if (config.subscriptionKey.isNotEmpty()) {
          SpeechConfig.fromEndpoint(uri, config.subscriptionKey)
        } else {
          SpeechConfig.fromEndpoint(uri)
        }.also { built ->
          if (config.authToken.isNotEmpty()) built.authorizationToken = config.authToken
        }
      }
      config.authToken.isNotEmpty() ->
        SpeechConfig.fromAuthorizationToken(config.authToken, config.region)
      else ->
        SpeechConfig.fromSubscription(config.subscriptionKey, config.region)
    }

    cfg.speechRecognitionLanguage = config.language

    val profanity = when (config.profanity.lowercase()) {
      "removed" -> ProfanityOption.Removed
      "raw", "none" -> ProfanityOption.Raw
      else -> ProfanityOption.Masked
    }
    cfg.setProfanity(profanity)

    if (config.outputFormat.lowercase() == "detailed") {
      cfg.outputFormat = OutputFormat.Detailed
      cfg.requestWordLevelTimestamps()
    }

    // Silence-detection: auto-stop after N ms of silence (server-side VAD).
    if (config.silenceTimeoutMs > 0) {
      cfg.setProperty(
        PropertyId.SpeechServiceConnection_EndSilenceTimeoutMs,
        config.silenceTimeoutMs.toString(),
      )
    }

    return cfg
  }

  /**
   * Install phrase-list grammar on the live recognizer for custom-vocab biasing.
   * No-op if no hints were configured. Must be called AFTER the recognizer
   * exists.
   */
  private fun installPhraseHints() {
    if (config.phraseHints.isEmpty()) return
    val rec = engine?.sdkRecognizer ?: return
    val grammar = PhraseListGrammar.fromRecognizer(rec)
    for (phrase in config.phraseHints) {
      if (phrase.isNotEmpty()) grammar.addPhrase(phrase)
    }
  }

  // MARK: AzureRecognizerCallbacks

  override fun azureRecognizing(result: RecognitionResult, speakerId: String) {
    if (!active) return  // Don't emit partials in the tail window.
    val transcript = buildTranscript(result, isFinal = false, speakerId = speakerId)
    latestPartial = transcript.text
    if (config.emitPartials) {
      callbacks.onPartial(transcript)
    }
  }

  override fun azureRecognized(result: RecognitionResult, speakerId: String) {
    if (!acceptingEvents()) return
    if (result.reason != ResultReason.RecognizedSpeech) return
    val transcript = buildTranscript(result, isFinal = true, speakerId = speakerId)
    if (transcript.text.isNotEmpty()) {
      accumulatedFinals += transcript.text
    }
    callbacks.onFinal(transcript)
  }

  /** Shared builder call so partial + final paths don't duplicate the same 8-line invocation. */
  private fun buildTranscript(
    result: RecognitionResult,
    isFinal: Boolean,
    speakerId: String,
  ): AzureTranscript = AzureTranscriptBuilder.from(
    result = result,
    isFinal = isFinal,
    sessionId = sessionId,
    sessionStartMs = sessionStartMs,
    fallbackLanguage = config.language,
    speakerId = speakerId,
  )

  override fun azureCanceled(
    reason: CancellationReason?,
    errorCode: CancellationErrorCode?,
    errorDetails: String?,
  ) {
    if (reason == CancellationReason.EndOfStream) return
    if (!acceptingEvents()) return
    val mapped = AzureErrorMapping.map(
      reason = reason ?: CancellationReason.Error,
      errorCode = errorCode ?: CancellationErrorCode.NoError,
      errorDetails = errorDetails,
    )
    callbacks.onError(mapped)
  }

  override fun azureSessionStopped() {
    // Close the tail window — recognizer has finished flushing.
    tailWindowEndMs = 0L
  }

  override fun azureSpeechBoundary(kind: String, offsetMs: Double) {
    if (!active) return
    callbacks.onSpeechBoundary(kind, offsetMs)
  }
}
